plugins/elantp/fu-elantp-hid-device.c

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <linux/hidraw.h>
+#include <linux/input.h>
+
+#include "fu-elantp-common.h"
+#include "fu-elantp-firmware.h"
+#include "fu-elantp-hid-device.h"
+
+#include "fu-chunk.h"
+
+struct _FuElantpHidDevice {
+	FuUdevDevice		 parent_instance;
+	guint16			 ic_page_count;
+	guint16			 iap_ctrl;
+	guint16			 module_id;
+};
+
+G_DEFINE_TYPE (FuElantpHidDevice, fu_elantp_hid_device, FU_TYPE_UDEV_DEVICE)
+
+
+static void
+fu_elantp_hid_device_to_string (FuDevice *device, guint idt, GString *str)
+{
+	FuElantpHidDevice *self = FU_ELANTP_HID_DEVICE (device);
+	fu_common_string_append_kx (str, idt, "ModuleId", self->module_id);
+	fu_common_string_append_kx (str, idt, "IcPageCount", self->ic_page_count);
+	fu_common_string_append_kx (str, idt, "EapCtrl", self->iap_ctrl);
+}
+
+static gboolean
+fu_elantp_hid_device_probe (FuUdevDevice *device, GError **error)
+{
+	/* check is valid */
+	if (g_strcmp0 (fu_udev_device_get_subsystem (device), "hidraw") != 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "is not correct subsystem=%s, expected hidraw",
+			     fu_udev_device_get_subsystem (device));
+		return FALSE;
+	}
+
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id (device, "hid", error);
+}
+
+static gboolean
+fu_elantp_hid_device_send_cmd (FuElantpHidDevice *self,
+			       guint8 *tx, gsize txsz,
+			       guint8 *rx, gsize rxsz,
+			       GError **error)
+{
+	g_autofree guint8 *buf = NULL;
+
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCSFEATURE(txsz), tx,
+				   NULL, error))
+		return FALSE;
+	if (rxsz == 0)
+		return TRUE;
+
+	/* GetFeature */
+	buf = g_malloc0 (rxsz + 1);
+	buf[0] = tx[0]; /* report number */
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(rxsz + 3), buf,
+				   NULL, error))
+		return FALSE;
+
+	/* success */
+	memcpy (rx, buf + 0x3, rxsz);
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_hid_device_read_cmd (FuElantpHidDevice *self, guint16 reg,
+			       guint8 *rx, gsize rxsz, GError **error)
+{
+	guint8 buf[5];
+
+	buf[0] = 0x0d; /* report number */
+	buf[1] = 0x05;
+	buf[2] = 0x03;
+	fu_common_write_uint16 (buf + 0x3, reg, G_LITTLE_ENDIAN);
+	if (g_getenv ("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "ReadCmd", buf, sizeof(buf));
+	return fu_elantp_hid_device_send_cmd (self, buf, sizeof(buf), rx, rxsz, error);
+}
+
+static gint
+fu_elantp_hid_device_write_cmd (FuElantpHidDevice *self,
+				guint16 reg, guint16 cmd,
+				GError **error)
+{
+	guint8 buf[5];
+
+	buf[0] = 0x0d; /* report number */
+	fu_common_write_uint16 (buf + 0x1, reg, G_LITTLE_ENDIAN);
+	fu_common_write_uint16 (buf + 0x3, cmd, G_LITTLE_ENDIAN);
```

Comment:
same here, don't you have one byte of garbage since you're only writing 2 bytes?